File: resources/vis-python/extension_check.py
"""Static check of a Python extension file: parse it, never import it.

The point is to answer "would this file load, and would the forms it asks for be
accepted" WITHOUT running a single line of it. So there is no import here and no
`eval` of the file: it is parsed with `ast`, and only the expressions that are
statically knowable are turned into values - literals, containers, and calls to
the `vis` form builders, which are pure dict constructors.

Three things come out of that:

* a syntax error, with its own line;
* every `vis.<name>` the module reads that the `vis` module does not have, which
  is how `vis.plaintxt(...)` is caught before it raises in front of a human;
* every `vis.ask(...)` whose request can be reconstructed, validated through
  the engine's own seam via `vis._check` - the same judge the running dialog
  uses, never a second opinion about it.

A `validate=` function is checked by SHAPE, never called: a lambda or a `def` in
the file stands in as a placeholder taking exactly the arguments it declares, so
`validate=lambda: None` is refused here for the same reason `vis.ask` refuses it.

An argument that cannot be known without running the file (a variable from an
argument, an f-string, a comprehension) is COUNTED as skipped rather than
guessed at: a checker that invents values would report problems nobody has.
"""

import ast
import json

import vis

# Pure node constructors: calling one builds a dict and touches nothing else,
# which is why a static check is allowed to call them.
_BUILDERS = (
    "plaintext",
    "password",
    "multiline",
    "select",
    "multiselect",
    "checkbox",
    "slider",
    "otp",
    "option",
    "row",
    "column",
    "heading",
    "paragraph",
)

# The one entry point that carries a request worth validating.
_ASKS = ("ask",)


class _Unknown(Exception):
    """This expression's value is not knowable without running the file."""


# Stand-ins for a `validate=` function. A validator is CODE, so it is never run;
# only its call shape matters, and these carry every shape the engine tells
# apart: the value alone, the value and every value, anything at all, and one
# that can take neither (which is exactly what `vis.ask` refuses).
def _one(_value):
    return None


def _two(_value, _values):
    return None


def _any_args(*_args):
    return None


def _refused():
    return None


def _placeholder(args):
    """A stand-in with the same call shape as this `lambda` / `def` signature."""
    if args.vararg is not None:
        return _any_args
    if any(default is None for default in args.kw_defaults):
        return _refused  # a required keyword-only argument: never fillable
    total = len(getattr(args, "posonlyargs", [])) + len(args.args)
    required = total - len(args.defaults)
    if total < 1 or required > 2:
        return _refused
    return _two if total >= 2 else _one


def _vis_names(tree):
    """Every local name bound to the `vis` module by an `import`."""
    names = set()
    for node in ast.walk(tree):
        if isinstance(node, ast.Import):
            for alias in node.names:
                if alias.name == "vis":
                    names.add(alias.asname or "vis")
    return names or {"vis"}


def _read(node, names):
    """The `<vis>.<attr>` this node reads, or None."""
    if (
        isinstance(node, ast.Attribute)
        and isinstance(node.value, ast.Name)
        and node.value.id in names
    ):
        return node.attr
    return None


def _value(node, env, names):
    """`node`'s value, or `_Unknown` when only running the file would tell."""
    if isinstance(node, ast.Constant):
        return node.value
    if isinstance(node, (ast.List, ast.Tuple, ast.Set)):
        return [_value(e, env, names) for e in node.elts]
    if isinstance(node, ast.Dict):
        return {
            _value(k, env, names): _value(v, env, names)
            for k, v in zip(node.keys, node.values)
        }
    if isinstance(node, ast.UnaryOp) and isinstance(node.op, ast.USub):
        return -_value(node.operand, env, names)
    if isinstance(node, ast.Name):
        if node.id in env:
            return env[node.id]
        raise _Unknown(node.id)
    if isinstance(node, ast.Lambda):
        return _placeholder(node.args)
    if isinstance(node, ast.Call):
        builder = _read(node.func, names)
        if builder in _BUILDERS:
            args = [_value(a, env, names) for a in node.args]
            return getattr(vis, builder)(*args, **_kwargs(node, env, names))
    raise _Unknown(type(node).__name__)


def _kwargs(node, env, names):
    """A call's keyword arguments, with an unknowable `validate=` stood in for."""
    out = {}
    for kw in node.keywords:
        if kw.arg is None:
            raise _Unknown("**kwargs")
        try:
            out[kw.arg] = _value(kw.value, env, names)
        except _Unknown:
            # A validator nobody can see IS code; the request only cares that it
            # is callable, so the common one-value shape stands in for it.
            if kw.arg != "validate":
                raise
            out[kw.arg] = _one
    return out


def _problem(kind, node, message):
    return {
        "kind": kind,
        "line": getattr(node, "lineno", 0),
        "column": getattr(node, "col_offset", 0) + 1,
        "message": message,
    }


def _request_args(node, env, names):
    """The `(title, fields, options)` a `vis.ask` call carries."""
    positional = [_value(a, env, names) for a in node.args]
    options = _kwargs(node, env, names)
    title = positional[0] if positional else options.pop("title")
    fields = positional[1] if len(positional) > 1 else options.pop("fields")
    return title, fields, options


def _check_ask(node, env, names, report):
    try:
        title, fields, options = _request_args(node, env, names)
    except (_Unknown, KeyError, IndexError):
        report["skipped"] += 1
        return
    report["checked"] += 1
    try:
        reason = vis._check(title, fields, **options)
    except Exception as exc:  # a builder refused the shape outright
        reason = str(exc)
    if reason:
        report["problems"].append(_problem("invalid-request", node, reason))


def _walk(node, env, names, report):
    """Depth first, in source order: bind what is knowable, check every ask."""
    for child in ast.iter_child_nodes(node):
        if isinstance(child, (ast.FunctionDef, ast.AsyncFunctionDef)):
            # `validate=is_a_slug` is knowable: the name stands for its shape.
            env[child.name] = _placeholder(child.args)
        if isinstance(child, ast.Attribute):
            attr = _read(child, names)
            if attr and not hasattr(vis, attr):
                report["problems"].append(
                    _problem(
                        "unknown-attribute",
                        child,
                        "the vis module has no " + attr,
                    )
                )
        if isinstance(child, ast.Call) and _read(child.func, names) in _ASKS:
            _check_ask(child, env, names, report)
        if (
            isinstance(child, ast.Assign)
            and len(child.targets) == 1
            and isinstance(child.targets[0], ast.Name)
        ):
            target = child.targets[0].id
            try:
                env[target] = _value(child.value, env, names)
            except _Unknown:
                env.pop(target, None)
        _walk(child, env, names, report)


def vis_check_source(source, path):
    """Check one extension file's `source`, reported as a JSON string."""
    report = {"path": path, "problems": [], "checked": 0, "skipped": 0}
    try:
        tree = ast.parse(source, filename=path)
    except SyntaxError as exc:
        report["problems"].append(
            {
                "kind": "syntax",
                "line": exc.lineno or 0,
                "column": exc.offset or 0,
                "message": exc.msg,
            }
        )
        report["is_valid"] = False
        return json.dumps(report)
    names = _vis_names(tree)
    _walk(tree, {}, names, report)
    if not any(
        isinstance(node, ast.Call) and _read(node.func, names) == "extension"
        for node in ast.walk(tree)
    ):
        report["problems"].append(
            {
                "kind": "no-extension",
                "line": 0,
                "column": 0,
                "message": "nothing calls vis.extension(...), so this file registers nothing",
            }
        )
    report["is_valid"] = not report["problems"]
    return json.dumps(report)

File: src/com/blockether/vis/internal/extension_check.clj
(ns ^:no-doc com.blockether.vis.internal.extension-check
  "Static check of a Python extension file: does it parse, does it only reach for
   `vis` names that exist, and would the forms it asks for be accepted?

   Running the file would answer all three, but running it is exactly what an
   author cannot afford before a human is watching: an extension's top level may
   shell out, write state or mount a dialog. So nothing here executes the
   extension. The file is PARSED (`vis-python/extension_check.py`, `ast` only),
   the statically knowable arguments of every `vis.ask` and every `vis.live` call
   are reconstructed, and what comes out is judged by CALLING
   [[com.blockether.vis.internal.view/normalize-request]] or
   [[com.blockether.vis.internal.view/normalize-live-view]] -- the engine's
   own normalizers, the seams every running dialog and every mounted view cross,
   so a refusal here is the line the author would have seen in front of the human.

   The `vis` module the checker reads is the REAL one: the bootstrap is evaluated
   with every host callback bound to a refusal (`python-extensions/bind-inert-host!`),
   so `vis.plaintext(...)` builds its dict while `vis.shell(...)` cannot run. The
   two exceptions are `__vis_host_request_input__`, bound to a judge that
   normalizes the request and settles it `undeliverable`, and `__vis_host_live__`,
   bound to a judge that normalizes the view and mounts nothing: a form is checked
   by ASKING for it and a view by OPENING it, and neither reaches a human.

   What cannot be known without running the file (a field list assembled from an
   argument, an f-string title, a comprehension) is COUNTED as skipped, never
   guessed at: a checker that invents values reports problems nobody has."
  (:require [charred.api :as json]
            [clojure.java.io :as io]
            [clojure.string :as str]
            [com.blockether.vis.internal.env-python :as env]
            [com.blockether.vis.internal.gateway.wire :as wire]
            [com.blockether.vis.internal.view :as view]
            [com.blockether.vis.internal.python-extensions :as px])
  (:import (java.io File)
           (org.graalvm.polyglot Context Value)))

(def ^:private checker-python
  "The checker's Python body, slurped from its classpath resource through the ONE
   shared reader so a resource missing from the native image fails loudly."
  (env/runtime-python-src "vis-python/extension_check.py"))

(def ^:private unasked-answer
  "What [[judge-request]] hands a checked `vis.ask` back: nobody was asked, so the
   only honest settlement is the engine's own word for a request that reached no
   surface."
  (json/write-json-str (view/answer->wire {:is-submitted false :reason "undeliverable"})))

(defn- judge-request
  "The checker's `request_input`: normalize the request the way a real ask does,
   then answer `undeliverable` instead of drawing, publishing or parking
   anything. A bad spec throws here exactly where `vis.ask` throws for an
   extension that runs, and that line is what the checker reports."
  [request-json & _]
  (let [request (json/read-json (str request-json) :key-fn identity)]
    (view/normalize-request (cond-> request
                              (map? request)
                              (dissoc "channel_id" "channel_ids")))
    unasked-answer))

(defn- judge-view
  "The checker's `live`: normalize the view the way a real `vis.live` does, then
   answer the handle it would have answered -- without mounting it, publishing an
   event or opening a record. A node the vocabulary has no name for throws here
   exactly where the engine throws for an extension that runs, and that line is
   what the checker reports."
  [envelope-json & _]
  (let [envelope
        (json/read-json (str envelope-json) :key-fn identity)

        declared
        (get envelope "view")

        view
        (view/normalize-live-view (cond-> declared
                                    (map? declared)
                                    (dissoc "channel_id" "channel_ids")))]

    (wire/json-str {:view-id (:id view) :is-open true :view view})))

(defn- checker-context
  "A context holding the real `vis` module with an inert host, plus the checker."
  ^Context []
  (let [^Context ctx (px/build-context "extension-check")]
    (px/bind-inert-host! ctx
                         {"__vis_host_request_input__" judge-request
                          "__vis_host_live__" judge-view})
    (locking ctx
      (.eval ctx "python" ^String px/bootstrap-python)
      (.eval ctx "python" ^String checker-python))
    ctx))

(defn- problem [kind line column message] {:kind kind :line line :column column :message message})

(defn- ->report
  "The checker's JSON verdict as a Clojure report."
  [path json-str]
  (let [m (json/read-json (str json-str) :key-fn identity)]
    {:path path
     :checked (long (get m "checked" 0))
     :skipped (long (get m "skipped" 0))
     :is-valid (boolean (get m "is_valid"))
     :problems (mapv (fn [p]
                       (problem (get p "kind")
                                (long (get p "line" 0))
                                (long (get p "column" 0))
                                (get p "message")))
                     (get m "problems"))}))

(defn- check-in
  "Check one `source` inside an already-built checker context."
  [^Context ctx source path]
  (locking ctx
    (let [f ^Value (.getMember (.getBindings ctx "python") "vis_check_source")]
      (->report path (.asString ^Value (.execute f (object-array [(str source) (str path)])))))))

(defn- unreadable
  [path message]
  {:path path :checked 0 :skipped 0 :is-valid false :problems [(problem "unreadable" 0 0 message)]})

(defn check-sources
  "Check a seq of `[path source]` pairs in ONE context. Returns one report per
   pair: `{:path :checked :skipped :is-valid :problems [{:kind :line :column
   :message}]}`. Reports are data -- a file with problems is a verdict, never a
   throw, so a run over a directory always reaches the last file."
  [pairs]
  (if-not (seq pairs)
    []
    (let [ctx (checker-context)]
      (try (mapv (fn [[path source]]
                   (check-in ctx source path))
                 pairs)
           (finally (.close ctx))))))

(defn check-source
  "Check one extension `source`, labelled `path`."
  ([source] (check-source source "<source>"))
  ([source path] (first (check-sources [[path source]]))))

(defn check-files
  "Check `files` (anything `io/file` accepts). A file that cannot be read is its
   own `unreadable` problem rather than an aborted run."
  [files]
  (let [read-one
        (fn [f]
          (let [^File file (io/file f)]
            (try [(.getPath file) (slurp file)] (catch Exception e [(.getPath file) e]))))

        pairs
        (mapv read-one files)

        ok
        (filterv (fn [[_ source]]
                   (string? source))
          pairs)

        reports
        (into {} (map (juxt :path identity)) (check-sources ok))]

    (mapv (fn [[path source]]
            (if (string? source)
              (get reports path)
              (unreadable path (ex-message ^Exception source))))
          pairs)))

(defn expand-paths
  "Every file `paths` names: a directory contributes its own `*.py` in name
   order, non-recursively. No paths at all means the extension directories vis
   itself loads, so `vis-agent extension check` with no arguments checks exactly
   what would load next."
  [paths]
  (let [roots (if (seq paths) (mapv io/file paths) (px/default-extension-dirs))]
    (vec (mapcat (fn [^File f]
                   (if (.isDirectory f)
                     (sort (filter (fn [^File c]
                                     (str/ends-with? (.getName c) ".py"))
                                   (or (seq (.listFiles f)) [])))
                     [f]))
                 roots))))

(defn ok? "Did every report come back valid?" [reports] (every? :is-valid reports))

(defn- plural [n word] (str n " " word (when (not= 1 n) "s")))

(defn- report-lines
  [{:keys [path problems checked skipped is-valid]}]
  (into [(str (if is-valid "ok   " "FAIL ")
              path
              "  ("
              (plural checked "form")
              " checked"
              (when (pos? ^long skipped) (str ", " skipped " skipped"))
              ")")]
        (map (fn [{:keys [kind line column message]}]
               (str "  " path ":" line ":" column ": " kind ": " message)))
        problems))

(defn report-text
  "The whole run as the text `vis-agent extension check` prints: one line per
   file, one indented line per problem, and a closing tally."
  [reports]
  (let [total
        (reduce + 0 (map (comp long :checked) reports))

        bad
        (reduce + 0 (map (comp count :problems) reports))]

    (str/join "\n"
              (concat (mapcat report-lines reports)
                      [(str (plural (count reports) "file")
                            ", " (plural total "form")
                            " checked, " (plural bad "problem"))]))))
